pkg/sync/service.go

```diff
@@ -388,6 +388,73 @@ func (s *Service) doDownloadFromURL(request *http.Request) (err error) {
 	return nil
 }
 
+func (s *Service) CloneFromBackingImage(writer http.ResponseWriter, request *http.Request) {
+	err := s.doCloneFromBackingImage(request)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (s *Service) doCloneFromBackingImage(request *http.Request) (err error) {
+	defer func() {
+		if err != nil {
+			s.log.WithError(err).Errorf("Sync Service: failed to do clone from backing image")
+		}
+	}()
+
+	queryParams := request.URL.Query()
+	sourceBackingImage := queryParams.Get(types.DataSourceTypeCloneParameterBackingImage)
+	if sourceBackingImage == "" {
+		return fmt.Errorf("%v is not specified", types.DataSourceTypeCloneParameterBackingImage)
+	}
+
+	sourceBackingImageUUID := queryParams.Get(types.DataSourceTypeCloneParameterBackingImageUUID)
+	if sourceBackingImageUUID == "" {
+		return fmt.Errorf("%v is not specified", types.DataSourceTypeCloneParameterBackingImageUUID)
+	}
+
+	encryption := types.EncryptionType(queryParams.Get(types.DataSourceTypeCloneParameterEncryption))
+	if encryption != types.EncryptionTypeEncrypt && encryption != types.EncryptionTypeDecrypt && encryption != types.EncryptionTypeIgnore {
+		return fmt.Errorf("%v operation %v is not specified", types.DataSourceTypeCloneParameterEncryption, encryption)
+	}
+	filePath := queryParams.Get("file-path")
+	if filePath == "" {
+		return fmt.Errorf("no filePath restoring file")
+	}
+	uuid := queryParams.Get("uuid")
+	if uuid == "" {
+		return fmt.Errorf("no uuid for restoring file")
+	}
+	diskUUID := queryParams.Get("disk-uuid")
+	expectedChecksum := queryParams.Get("expected-checksum")
```

Comment:
Is `expected-checksum"` available for decryption or encryption?

Reply:
no, we won't be able to know the checksum before hand
especially for encryption, there is 16MB metadata added by LUKS

`expected-checksum` is added by users when creating backing image. I don't think they know the checksum as well.
But we should keep it here to make the interface align

Reply:
So, is `expectedChecksum` always an empty string here?

Reply:
clone + ignore can use `expectedChecksum`